### PR TITLE
adding FT news channel

### PR DESCRIPTION
**Pull request by dcstang**
Adding a channel that I use on a daily basis!

Title - "Financial Times News Briefing"
Shortcode - "FT"
Feed URL (cant be used directly) - "https://www.ft.com/newsbriefing"
Alternative names that may be used by a user - "FT News"
Country - United Kingdom (UK)

Unfortunately the mp3 link is hidden behind a second page, and there isn't any RSS feed. Therefore, I resorted to using bs4. The date module is to match the latest news, as bs4 doesn't parse the page in top to bottom fashion. Let me know your thoughts, thanks!

### CODE REVIEW

**Comment by forslund** [comment]
Hi, always annoying when you have to resort to scraping for these things.

The first comment I have is that bs4 need to be added to the requirements.txt. Will look at this some more. Thanks for contributing

**Comment by krisgesling** [comment]
This function is working great for me.

To let people trigger it, we then need to add the channel into the `FEEDS` dictionary around line 136:
```
    'FT': ('Financial Times', ft, None),
```
This gives the channel a name, points to your custom function, and says we don't currently have an image to display on GUI screens.

I also thought it would be good to include an alias under `dialog/en-us/alt.feed.name.value`:
```
financial,FT
```
This will let people say things like "play the financial news" rather than the full title.

My only other comment is that we can merge your datetime import with the line above becoming:
```
from datetime import date, timedelta
```

**Comment by forslund** [comment]
A minor code style thing is that we don't use camelCase in our code (other than for class names per PEP8), there are a couple of those that should be change (mp3Page -> mp3_page)

**Comment by dcstang** [comment]
Thank you for the feedback and review! Appreciate the input about PEP8 - have changed it! Added code to `FEEDS`, the `alt.feed.name.value`, and tidied-up `timedate` as mentioned by Kris.

**Comment by dcstang** [comment]
using datetime Object instead of string comparison, to prevent breaking
also added 2 spaces after function as per PEP8

**Comment by krisgesling** [comment]
Hey, sorry this dropped off my radar, but I wanted to come back and get it merged.

I just noticed that as I'm in Australia, I am sometimes _in the future_, ie "tomorrows date" (from a US perspective). It therefore can't find a link for my date. 

Perhaps we should accept today or yesterdays date?
```
    # Check if div matches today's or yesterday's date
    div_date = datetime.strptime(result.contents[0], '%A, %d %B, %Y').date()
    today = datetime.date.today()
    yesterday = today - timedelta(1)

    if div_date in (today, yesterday):
        target_div = result.parent.find_next('div')
```
Or check for one and if not found, look for yesterdays?

**Comment by dcstang** [comment]
No worries, let me add that in. Good point about the timezones.

**Comment by dcstang** [comment]
All done Kris! Let me know if you need anything else.
Edit: just rebased from the upstream branch to sync up with all the recent edits.

File: __init__.py [hunk 1]
```diff
@@ -78,6 +80,32 @@ def gbp():
...
+    """Custom news fetcher for today's FT news briefing"""
+    url = 'https://www.ft.com/newsbriefing'
+    page = urlopen(url)
+
+    # Use bs4 to parse website and get mp3 link
+    soup = BeautifulSoup(page, features='html.parser')
+    result = soup.find('time')
+
+    # Get today and yesterday's date to match div and play most recent news
+    today = datetime.date.today()
```
Comment (krisgesling) [inline __init__.py]: Hey thanks for updating and rebasing.

In my previous comment I didn't take into account that we were already importing datetime from datetime which means this is trying to access `datetime.datetime.date.today()`
So just need to switch this to:
```
today = date.today()
```
and then add `date` to the import at the top:
```
from datetime import date, datetime, timedelta
```

Comment (dcstang) [reply __init__.py]: My bad, that's all done up now!

**Comment by krisgesling** [comment]
Sorry, I just ran into another issue :(

They only do briefings on weekdays, so if a user requests this station on a Sunday (or Monday if you're down under) it will return None and throw an error. 

I'm not overly familiar with BS4, but am I correct in believing that `soup.find(element)` will always return the first instance of that element on the page?

If we are reasonably confident that:
- the element selector will only reference briefing episodes
- the briefing episodes will always be sorted in chronological order, hence the first in the list will always be the latest episode

Then the simplest solution may be to remove the date check altogether? This would cut it back to a much simpler:
```
def ft():
    """Custom news fetcher for today's FT news briefing"""
    url = 'https://www.ft.com/newsbriefing'
    page = urlopen(url)

    # Use bs4 to parse website and get mp3 link
    soup = BeautifulSoup(page, features='html.parser')
    result = soup.find('time')
    target_div = result.parent.find_next('div')
    target_url = 'http://www.ft.com' + target_div.a['href']
    mp3_page = urlopen(target_url)
    mp3_soup = BeautifulSoup(mp3_page, features='html.parser')
    mp3_url = mp3_soup.find('source')['src']

    return mp3_url
```
What do you think?

**Comment by krisgesling** [comment]
Alternatively we'd need to do some slightly fancy date checking to find the most recent US weekday

**Comment by dcstang** [comment]
Hi Kris, thanks for checking about the weekends. This is important, as Financial Times only releases news from Monday-Friday, so running this on a weekend will return None. Let me test out over the next 2 weekends to make sure everything is working spiffy before pushing further code. I'm in for a simpler solution, will need to match it with the div layout of FT. 

Edit: sorry for the delay in reply!

**Comment by krisgesling** [comment]
Closed by #110

**Pull request closed without merging**